pkg/oidc/device_auth.go

```diff
@@ -0,0 +1,229 @@
+	DeviceAuthGrantType = "urn:ietf:params:oauth:grant-type:device_code"
+
+	nameClaimKey = "name"
+)
+
+type DeviceAuthenticator struct {
+	clientID string
+	authURL  string
+}
+
+type DeviceCodeResponseData struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationUri         string `json:"verification_uri"`
+	VerificationUriComplete string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	Interval                int    `json:"interval"`
+}
+
+type DeviceTokenResponseData struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	ExpiresIn    int    `json:"expires_in"`
+	RefreshToken string `json:"refresh_token"`
+	IdToken      string `json:"id_token"`
+}
+
+type ErrorResponseData struct {
+	Error string `json:"error"`
+}
+
+type JwtName struct {
+	Name string `json:"name"`
+}
+
+func NewDeviceAuthenticator(clientID string, i *installation.Installation) *DeviceAuthenticator {
+	return &DeviceAuthenticator{
+		clientID: clientID,
+		authURL:  i.AuthURL,
+	}
+}
+
+func (a *DeviceAuthenticator) LoadDeviceCode() (DeviceCodeResponseData, error) {
+	formData := url.Values{}
+	formData.Add(DeviceAuthKeyClientID, a.clientID)
+	formData.Add(DeviceAuthKeyScope, DeviceAuthScopes)
+
+	result := DeviceCodeResponseData{}
+	response, err := http.PostForm(fmt.Sprintf(deviceCodeUrlTemplate, a.authURL), formData)
+	if err != nil {
+		return result, microerror.Maskf(cannotGetDeviceCodeError, err.Error())
+	}
+
+	responseBytes, err := bytesFromResponse(response)
+	if err != nil {
+		return result, microerror.Maskf(cannotGetDeviceCodeError, err.Error())
+	}
+
+	if err != nil {
+		return result, microerror.Maskf(cannotGetDeviceCodeError, err.Error())
+	}
+
+	err = json.Unmarshal(responseBytes, &result)
+	if err != nil {
+		return result, microerror.Maskf(cannotGetDeviceCodeError, err.Error())
+	}
+
+	return result, nil
+}
+
+func (a *DeviceAuthenticator) AwaitDeviceToken(data DeviceCodeResponseData) (DeviceTokenResponseData, string, error) {
+	loadTokenTicker := time.NewTicker(time.Duration(data.Interval*1000+250) * time.Millisecond)
+	expirationTimer := time.NewTimer(time.Duration(data.ExpiresIn) * time.Second)
+	responseCh := make(chan DeviceTokenResponseData)
+	errCh := make(chan error)
+
+	go func() {
+		for {
+			<-loadTokenTicker.C
+			response, err := loadDeviceToken(a.authURL, data.DeviceCode)
+			if err == nil {
+				responseCh <- response
+				return
+			} else if !IsAuthorizationPendingError(err) && !IsTooManyAuthRequestsError(err) {
+				errCh <- err
+				return
+			}
+		}
+	}()
+
+	go func() {
+		<-expirationTimer.C
+		errCh <- cannotGetDeviceTokenError
+	}()
+
+	var response DeviceTokenResponseData
+	var err error
+	select {
+	case response = <-responseCh:
+		break
+	case err = <-errCh:
+		break
+	}
+
+	expirationTimer.Stop()
+	loadTokenTicker.Stop()
```

Comment:
Isn't it safer to defer right after start?

Reply:
I think it is very impoetant to stop the timer and ticker as soon as the channels are read.
With defer they would be stopped after the JWT is processed, which may be too late.
I moved it to a separate function and used defer.